File: backend/models/TranslationSuggestion.js
const mongoose = require('mongoose');

const translationSuggestionSchema = new mongoose.Schema({
  // Informations sur la plante
  scientificName: {
    type: String,
    required: true,
    trim: true
  },
  
  // Propositions de traduction
  darijaProposal: {
    type: String,
    trim: true,
    default: null
  },
  tamazightProposal: {
    type: String,
    trim: true,
    default: null
  },
  
  // Informations du contributeur
  contributorName: {
    type: String,
    required: true,
    trim: true
  },
  contributorEmail: {
    type: String,
    required: true,
    trim: true,
    lowercase: true
  },
  contributorRegion: {
    type: String,
    trim: true,
    default: ''
  },
  
  // Notes additionnelles
  notes: {
    type: String,
    trim: true,
    default: ''
  },
  
  // Statut de la proposition
  status: {
    type: String,
    enum: ['pending', 'approved', 'rejected', 'needs_review'],
    default: 'pending'
  },
  
  // Métadonnées
  submittedAt: {
    type: Date,
    default: Date.now
  },
  reviewedAt: {
    type: Date,
    default: null
  },
  reviewedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null
  },
  reviewNotes: {
    type: String,
    trim: true,
    default: ''
  },
  
  // Source de la proposition
  source: {
    type: String,
    enum: ['mobile_app', 'web_app', 'admin'],
    default: 'mobile_app'
  },
  
  // Validation
  isValidated: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true // Ajoute createdAt et updatedAt automatiquement
});

// Index pour la recherche rapide
translationSuggestionSchema.index({ scientificName: 1, status: 1 });
translationSuggestionSchema.index({ contributorEmail: 1 });
translationSuggestionSchema.index({ submittedAt: -1 });
translationSuggestionSchema.index({ status: 1, submittedAt: -1 });

// Validation : au moins une proposition doit exister
translationSuggestionSchema.pre('save', async function() {
  if (!this.darijaProposal && !this.tamazightProposal) {
    const error = new Error('Au moins une proposition (Darija ou Tamazight) est requise');
    throw error;
  }
});

// Méthode virtuelle pour vérifier si la proposition est valide
translationSuggestionSchema.virtual('hasValidProposal').get(function() {
  return (this.darijaProposal && this.darijaProposal.trim().length > 0) ||
         (this.tamazightProposal && this.tamazightProposal.trim().length > 0);
});

// Méthode pour approuver une proposition
translationSuggestionSchema.methods.approve = function(adminId, reviewNotes = '') {
  this.status = 'approved';
  this.reviewedAt = new Date();
  this.reviewedBy = adminId;
  this.reviewNotes = reviewNotes;
  this.isValidated = true;
  return this.save();
};

// Méthode pour rejeter une proposition
translationSuggestionSchema.methods.reject = function(adminId, reviewNotes = '') {
  this.status = 'rejected';
  this.reviewedAt = new Date();
  this.reviewedBy = adminId;
  this.reviewNotes = reviewNotes;
  this.isValidated = false;
  return this.save();
};

// Méthode pour marquer comme nécessitant une review
translationSuggestionSchema.methods.requestReview = function(adminId, reviewNotes = '') {
  this.status = 'needs_review';
  this.reviewedAt = new Date();
  this.reviewedBy = adminId;
  this.reviewNotes = reviewNotes;
  return this.save();
};

// Méthodes statiques pour les statistiques
translationSuggestionSchema.statics.getStats = function() {
  return this.aggregate([
    {
      $group: {
        _id: '$status',
        count: { $sum: 1 }
      }
    }
  ]);
};

translationSuggestionSchema.statics.getStatsByDateRange = function(startDate, endDate) {
  return this.aggregate([
    {
      $match: {
        submittedAt: {
          $gte: startDate,
          $lte: endDate
        }
      }
    },
    {
      $group: {
        _id: '$status',
        count: { $sum: 1 }
      }
    }
  ]);
};

// Transformer en JSON pour l'API
translationSuggestionSchema.methods.toJSON = function() {
  const suggestion = this.toObject();
  delete suggestion.__v;
  return suggestion;
};

module.exports = mongoose.model('TranslationSuggestion', translationSuggestionSchema);

File: frontend/lib/models/translation_suggestion.dart
class TranslationSuggestion {
  final String id;
  final String scientificName;
  final String? darijaProposal;
  final String? tamazightProposal;
  final String contributorName;
  final String contributorEmail;
  final String region;
  final String notes;
  final DateTime submittedAt;
  final ProposalStatus status;
  final DateTime? reviewedAt;
  final String? reviewedBy;
  final String? reviewNotes;
  final bool isValidated;

  TranslationSuggestion({
    required this.id,
    required this.scientificName,
    this.darijaProposal,
    this.tamazightProposal,
    required this.contributorName,
    required this.contributorEmail,
    this.region = '',
    this.notes = '',
    required this.submittedAt,
    this.status = ProposalStatus.pending,
    this.reviewedAt,
    this.reviewedBy,
    this.reviewNotes = '',
    this.isValidated = false,
  });

  factory TranslationSuggestion.fromJson(Map<String, dynamic> json) {
    return TranslationSuggestion(
      id: json['_id'] ?? json['id'] ?? '',
      scientificName: json['scientificName'] ?? '',
      darijaProposal: json['darijaProposal'] as String?,
      tamazightProposal: json['tamazightProposal'] as String?,
      contributorName: json['contributorName'] ?? '',
      contributorEmail: json['contributorEmail'] ?? '',
      region: json['contributorRegion'] ?? json['region'] ?? '',
      notes: json['notes'] ?? '',
      submittedAt: json['submittedAt'] != null
          ? DateTime.parse(json['submittedAt'])
          : DateTime.now(),
      isValidated: json['isValidated'] ?? false,
    );
  }

  static ProposalStatus _parseStatus(String? status) {
    switch (status) {
      case 'approved':
        return ProposalStatus.approved;
      case 'rejected':
        return ProposalStatus.rejected;
      case 'needs_review':
        return ProposalStatus.needsReview;
      default:
        return ProposalStatus.pending;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'scientificName': scientificName,
      'darijaProposal': darijaProposal,
      'tamazightProposal': tamazightProposal,
      'contributorName': contributorName,
      'contributorEmail': contributorEmail,
      'region': region,
      'notes': notes,
      'submittedAt': submittedAt.toIso8601String(),
      'status': status.toString().split('.').last,
      'reviewedAt': reviewedAt?.toIso8601String(),
      'reviewedBy': reviewedBy,
      'reviewNotes': reviewNotes,
      'isValidated': isValidated,
    };
  }

  TranslationSuggestion copyWith({
    String? id,
    String? scientificName,
    String? darijaProposal,
    String? tamazightProposal,
    String? contributorName,
    String? contributorEmail,
    String? region,
    String? notes,
    DateTime? submittedAt,
    ProposalStatus? status,
    DateTime? reviewedAt,
    String? reviewedBy,
    String? reviewNotes,
    bool? isValidated,
  }) {
    return TranslationSuggestion(
      id: id ?? this.id,
      scientificName: scientificName ?? this.scientificName,
      darijaProposal: darijaProposal ?? this.darijaProposal,
      tamazightProposal: tamazightProposal ?? this.tamazightProposal,
      contributorName: contributorName ?? this.contributorName,
      contributorEmail: contributorEmail ?? this.contributorEmail,
      region: region ?? this.region,
      notes: notes ?? this.notes,
      submittedAt: submittedAt ?? this.submittedAt,
      status: status ?? this.status,
      reviewedAt: reviewedAt ?? this.reviewedAt,
      reviewedBy: reviewedBy ?? this.reviewedBy,
      reviewNotes: reviewNotes ?? this.reviewNotes,
      isValidated: isValidated ?? this.isValidated,
    );
  }
}

enum ProposalStatus {
  pending,
  approved,
  rejected,
  needsReview;

  @override
  String toString() {
    switch (this) {
      case ProposalStatus.pending:
        return 'pending';
      case ProposalStatus.approved:
        return 'approved';
      case ProposalStatus.rejected:
        return 'rejected';
      case ProposalStatus.needsReview:
        return 'needs_review';
    }
  }
}
